AutoBlocker: Add major version upgrade blocker

ndelangen opened the pull request:
## What I did

This commit introduces a new blocker mechanism to prevent users from skipping major version upgrades in Storybook. The implementation includes:

- A new module to check and block upgrades with large major version gaps
- Comprehensive test coverage for version comparison logic
- Integration with existing autoblock system

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

This is going to be rather hard to manual QA, honestly I could use some tips.

This upgrade blocker is explicitly skipped when upgrading from and to canary releases and release-candidates.
So it can ONLY be tested with latest releases; and it will only block the upgrade if the major version difference is more than 1.

I added unit tests to ensure that comparison code is okay, and I left plenty of error handling code, so that the most common failure is to allow the upgrade.

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  79.7 MB | 79.7 MB | 0 B | **1.66** | 0% |
| initSize |  79.7 MB | 79.7 MB | 0 B | **1.66** | 0% |
| diffSize |  97 B | 97 B | 0 B | - | 0% |
| buildSize |  7.3 MB | 7.3 MB | 0 B | -0.47 | 0% |
| buildSbAddonsSize |  1.9 MB | 1.9 MB | 0 B | 0.44 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.88 MB | 1.88 MB | 0 B | -0.89 | 0% |
| buildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.97 MB | 3.97 MB | 0 B | -1.15 | 0% |
| buildPreviewSize |  3.33 MB | 3.33 MB | 0 B | 0.07 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |



| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  7.6s | 7.1s | -495ms | -0.98 | -6.9% |
| generateTime |  20.2s | 19.3s | -953ms | -0.31 | -4.9% |
| initTime |  5.1s | 4.6s | -476ms | -0.27 | -10.3% |
| buildTime |  8.6s | 9.4s | 756ms | -0.22 | 8% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  4.8s | 5.6s | 870ms | 0.51 | 15.3% |
| devManagerResponsive |  4.6s | 5.3s | 763ms | 0.46 | 14.1% |
| devManagerHeaderVisible |  748ms | 782ms | 34ms | -0.35 | 4.3% |
| devManagerIndexVisible |  760ms | 795ms | 35ms | -0.5 | 4.4% |
| devStoryVisibleUncached |  1.9s | 2s | 108ms | 0.48 | 5.4% |
| devStoryVisible |  780ms | 862ms | 82ms | -0.33 | 9.5% |
| devAutodocsVisible |  692ms | 774ms | 82ms | -0.05 | 10.6% |
| devMDXVisible |  702ms | 825ms | 123ms | 0.4 | 14.9% |
| buildManagerHeaderVisible |  646ms | 922ms | 276ms | **1.51** | 🔺29.9% |
| buildManagerIndexVisible |  674ms | 1s | 343ms | **1.82** | 🔺33.7% |
| buildStoryVisible |  614ms | 859ms | 245ms | **1.32** | 🔺28.5% |
| buildAutodocsVisible |  686ms | 584ms | -102ms | -0.64 | -17.5% |
| buildMDXVisible |  526ms | 582ms | 56ms | -0.28 | 9.6% |

<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

This PR adds a major version upgrade blocker to prevent users from skipping versions (e.g., v6 to v8), ensuring smoother migration paths between Storybook releases.

- Added `code/lib/cli-storybook/src/autoblock/block-major-version.ts` implementing version gap detection with special handling for prereleases and canary versions
- Added comprehensive test suite in `block-major-version.test.ts` covering various version scenarios and edge cases
- Modified `code/lib/cli-storybook/src/autoblock/index.ts` to integrate the new blocker into the existing autoblock system
- Implemented fail-safe error handling to allow upgrades when version detection fails
- Included helpful error messages with upgrade instructions when blocking occurs



<!-- /greptile_comment -->

ndelangen committed:
feat(cli): add major version upgrade blocker

This commit introduces a new blocker mechanism to prevent users from skipping major version upgrades in Storybook. The implementation includes:

- A new module to check and block upgrades with large major version gaps
- Comprehensive test coverage for version comparison logic
- Integration with existing autoblock system

ndelangen commented on code/lib/cli-storybook/src/autoblock/block-major-version.ts:
```diff
@@ -0,0 +1,93 @@
+    return false;
+  }
+
+  const gap = target.major - current.major;
+  return gap > 1;
```
downgrading makes no sense, and is unsupported

ndelangen committed:
refactor(cli): enhance major version upgrade blocking logic

Improves version upgrade checking with more granular status reporting:
- Replaces boolean check with enum-like result ('ok', 'downgrade', 'gap-too-large')
- Adds explicit handling for downgrade scenarios
- Expands test coverage to validate new upgrade checking behavior
- Updates blocker to return more detailed version mismatch information

ndelangen committed:
test(cli): expand version downgrade detection tests

Enhances test coverage for version downgrade scenarios by:
- Adding separate test suites for major, minor, and patch version downgrades
- Improving granularity of version comparison test cases

ndelangen committed:
Merge branch 'norbert/delete-deprecated' into norbert/upgrade-autoblocker

valentinpalkovic commented on code/lib/cli-storybook/src/autoblock/block-major-version.ts:
```diff
@@ -0,0 +1,115 @@
+        Downgrading is not supported.
+
+        For more information about upgrading and version compatibility, visit:
+        ${picocolors.cyan('https://storybook.js.org/docs/react/configure/upgrading')}`;
```
```suggestion
        ${picocolors.cyan('https://storybook.js.org/docs/configure/upgrading')}`;
```

valentinpalkovic commented on code/lib/cli-storybook/src/autoblock/block-major-version.ts:
```diff
@@ -0,0 +1,115 @@
+        ${picocolors.cyan(cmd)}
+
+        For more information about upgrading, visit:
+        ${picocolors.cyan('https://storybook.js.org/docs/react/configure/upgrading')}`;
```
```suggestion
        ${picocolors.cyan('https://storybook.js.org/docs/configure/upgrading')}`;
```

valentinpalkovic commented on code/lib/cli-storybook/src/autoblock/block-major-version.ts:
```diff
@@ -0,0 +1,115 @@
+      ${message}
+
+      For more information about upgrading, visit:
+      ${picocolors.cyan('https://storybook.js.org/docs/react/configure/upgrading')}`;
```
```suggestion
      ${picocolors.cyan('https://storybook.js.org/docs/configure/upgrading')}`;
```

ndelangen committed:
Merge branch 'next' into norbert/upgrade-autoblocker

ndelangen committed:
Apply suggestions from code review

Co-authored-by: Valentin Palkovic <valentin@chromatic.com>

ndelangen committed:
test(cli): update block-major-version tests with improved mocking

Refactors test suite for major version blocking with:
- Enhanced mocking of Storybook version retrieval
- Updated test cases for version downgrade scenarios
- Improved error handling and package JSON retrieval tests

ndelangen committed:
Merge branch 'norbert/upgrade-autoblocker' of https://github.com/storybookjs/storybook into norbert/upgrade-autoblocker

ndelangen committed:
Merge branch 'next' into norbert/upgrade-autoblocker
